deps/rabbit/src/rabbit_queue_index.erl

```diff
@@ -1215,11 +1217,11 @@ segment_plus_journal1(undefined, {?PUB, del, no_ack} = Obj) ->
 segment_plus_journal1(undefined, {?PUB, del, ack}) ->
     {undefined, 0};
 
-segment_plus_journal1({?PUB = Pub, no_del, no_ack}, {no_pub, del, no_ack}) ->
+segment_plus_journal1({?PUB = Pub, no_del, no_ack}, {no_pub, _del, no_ack}) ->
     {{Pub, del, no_ack}, 0};
-segment_plus_journal1({?PUB, no_del, no_ack},       {no_pub, del, ack}) ->
+segment_plus_journal1({?PUB, no_del, no_ack},       {no_pub, _del, ack}) ->
     {undefined, -1};
-segment_plus_journal1({?PUB, del, no_ack},          {no_pub, no_del, ack}) ->
+segment_plus_journal1({?PUB, del, no_ack},          {no_pub, _no_del, ack}) ->
     {undefined, -1}.
```

Comment:
User report that the following can happen on recovery:

```
{:failed_to_start_child, :rabbit_vhost_process, {:error, {{{:function_clause, [{:rabbit_queue_index, :journal_minus_segment1, [{:no_pub, :no_del, :ack}, {{true, <<...>>, <<...>>}, :no_del, :no_ack}], [file: 'src/rabbit_queue_index.erl', line: 1257]}, {:rabbit_queue_index, :"-journal_minus_segment/3-fun-0-", 4, [file: 'src/rabbit_queue_index.erl', line: 1234]}, {:array, :sparse_foldl_3, 7, [file: 'array.erl', line: 1684]}, {:array, :sparse_foldl_2, 9, [file: 'array.erl', line: 1678]}, {:rabbit_queue_index, :"-recover_journal/1-fun-0-", 1
```

It sounds like the segment was written but the journal remained on disk instead of being wiped, which may mean a node shut down abnormally in between the two.

It might be worth fixing that as part of this work or soon after. At the very least we don't want to prevent the vhost from starting if only one entry is wrong, as the fix would be to lose ALL messages. Food for thoughts.

Reply:
We've also seen the `:failed_to_start_child, :rabbit_vhost_process`-issue as per this gist: https://gist.github.com/johanrhodin/18c7415b88d56f8d32d7f1b7daa23a19 in RabbitMQ 3.8.19.
In that case we used strace to find the "faulty" .idx file and removed it, after that the vhost could start so not all messages were lost. If of interest we can share the data folder.

(I realize this is not a bug tracking discussion forum, but thought it was relevant. Let me know where to report this if it is of interest.)

Reply:
Is this after an abnormal shutdown as well?

Reply:
Not really abnormal, it was after an Erlang upgrade in place:

```
2021-07-27 10:40:52.454 [info] <0.282.0>
 Starting RabbitMQ 3.8.19 on Erlang 23.3.1 [emu]
```

Worked OK

```
2021-07-27 10:47:06.659 [info] <0.274.0>
 Starting RabbitMQ 3.8.19 on Erlang 24.0.2 [jit]
```

Followed by:

```
2021-07-27 10:47:07.901 [info] <0.599.0> Message store "B24YDZSK6ODW52OLV52U91PJQ/msg_store_transient": using rabbit_msg_store_ets_index to provide index
2021-07-27 10:47:07.903 [info] <0.595.0> Started message store of type transient for vhost 'VHOST1'
2021-07-27 10:47:07.903 [info] <0.603.0> Message store "B24YDZSK6ODW52OLV52U91PJQ/msg_store_persistent": using rabbit_msg_store_ets_index to provide index
2021-07-27 10:47:07.905 [warning] <0.603.0> Message store "B24YDZSK6ODW52OLV52U91PJQ/msg_store_persistent" : recovery terms differ from present
rebuilding indices from scratch
2021-07-27 10:47:07.908 [error] <0.447.0> ** Generic server <0.447.0> terminating
** Last message in was {'$gen_cast',{submit_async,#Fun<rabbit_queue_index.33.83626758>,<0.446.0>}}
```

Since it logs "recovery terms differ from present rebuilding indices from scratch" it must have detected some change?

Reply:
Sounds like the terms differ between what the vhost terms file says and what the message store terms file says. I don't think the data files would help me much since for the error it is mostly a matter of picking the most recent data. I'm wondering how you stopped RabbitMQ before upgrading though.

Reply:
> I'm wondering how you stopped RabbitMQ before upgrading though.

It is a regular `service rabbitmq-server stop`
```
Jul 27 10:40:33  rabbitmqctl[24821]: Shutting down RabbitMQ node rabbit@HOST1 running at PID 20214
Jul 27 10:40:34  rabbitmq-server[20214]: Gracefully halting Erlang VM
Jul 27 10:40:34  rabbitmq-server[20214]: Logger - error: {removed_failing_handler,rabbit_log}
Jul 27 10:40:34  rabbitmqctl[24821]: Waiting for PID 20214 to terminate
Jul 27 10:40:41  rabbitmqctl[24821]: RabbitMQ node rabbit@HOST1 running at PID 20214 successfully shut down
Jul 27 10:40:47  rabbitmq-server[25119]: Configuring logger redirection
Jul 27 10:40:52  rabbitmq-server[25119]: ## ## RabbitMQ 3.8.19
Jul 27 10:40:52  rabbitmq-server[25119]: ## ##
Jul 27 10:40:52  rabbitmq-server[25119]: ########## Copyright (c) 2007-2021 VMware, Inc. or its affiliates.
Jul 27 10:40:52  rabbitmq-server[25119]: ###### ##
Jul 27 10:40:52  rabbitmq-server[25119]: ########## Licensed under the MPL 2.0. Website: https://rabbitmq.com
Jul 27 10:40:52  rabbitmq-server[25119]: Erlang: 23.3.1 [emu]
Jul 27 10:40:52  rabbitmq-server[25119]: TLS Library: OpenSSL - OpenSSL 1.1.1 11 Sep 2018
Jul 27 10:40:52  rabbitmq-server[25119]: Doc guides: https://rabbitmq.com/documentation.html
Jul 27 10:40:52  rabbitmq-server[25119]: Support: https://rabbitmq.com/contact.html
Jul 27 10:40:52  rabbitmq-server[25119]: Tutorials: https://rabbitmq.com/getstarted.html
Jul 27 10:40:52  rabbitmq-server[25119]: Monitoring: https://rabbitmq.com/monitoring.html
Jul 27 10:40:52  rabbitmq-server[25119]: Logs: /var/log/rabbitmq/rabbit@HOST1.log
Jul 27 10:40:52  rabbitmq-server[25119]: /var/log/rabbitmq/rabbit@HOST1_upgrade.log
Jul 27 10:40:52  rabbitmq-server[25119]: Config file(s): /etc/rabbitmq/rabbitmq.config
Jul 27 10:40:54  rabbitmq-server[25119]: Starting broker... completed with 9 plugins.
Jul 27 10:46:03  rabbitmqctl[26447]: Shutting down RabbitMQ node rabbit@HOST1 running at PID 25119
Jul 27 10:46:04  rabbitmq-server[25119]: Logger - error: {removed_failing_handler,rabbit_log}
Jul 27 10:46:04  rabbitmq-server[25119]: Gracefully halting Erlang VM
Jul 27 10:46:04  rabbitmqctl[26447]: Waiting for PID 25119 to terminate
Jul 27 10:46:11  rabbitmqctl[26447]: RabbitMQ node rabbit@HOST1 running at PID 25119 successfully shut down
Jul 27 10:47:00  rabbitmq-server[27940]: Configuring logger redirection
Jul 27 10:47:06  rabbitmq-server[27940]: ## ## RabbitMQ 3.8.19
Jul 27 10:47:06  rabbitmq-server[27940]: ## ##
Jul 27 10:47:06  rabbitmq-server[27940]: ########## Copyright (c) 2007-2021 VMware, Inc. or its affiliates.
Jul 27 10:47:06  rabbitmq-server[27940]: ###### ##
Jul 27 10:47:06  rabbitmq-server[27940]: ########## Licensed under the MPL 2.0. Website: https://rabbitmq.com
Jul 27 10:47:06  rabbitmq-server[27940]: Erlang: 24.0.2 [jit]
Jul 27 10:47:06  rabbitmq-server[27940]: TLS Library: OpenSSL - OpenSSL 1.1.1 11 Sep 2018
Jul 27 10:47:06  rabbitmq-server[27940]: Doc guides: https://rabbitmq.com/documentation.html
Jul 27 10:47:06  rabbitmq-server[27940]: Support: https://rabbitmq.com/contact.html
Jul 27 10:47:06  rabbitmq-server[27940]: Tutorials: https://rabbitmq.com/getstarted.html
Jul 27 10:47:06  rabbitmq-server[27940]: Monitoring: https://rabbitmq.com/monitoring.html
Jul 27 10:47:06  rabbitmq-server[27940]: Logs: /var/log/rabbitmq/rabbit@HOST1.log
Jul 27 10:47:06  rabbitmq-server[27940]: /var/log/rabbitmq/rabbit@HOST1_upgrade.log
Jul 27 10:47:06  rabbitmq-server[27940]: Config file(s): /etc/rabbitmq/rabbitmq.config
Jul 27 10:47:08  rabbitmq-server[27940]: Starting broker...Logger - error: {removed_failing_handler,rabbit_log}
Jul 27 10:47:08  rabbitmq-server[27940]: completed with 9 plugins.
```
Corresponding RabbitMQ log:
```
2021-07-27 10:46:04.427 [info] <0.716.0> Stopping message store for directory '/var/lib/rabbitmq/mnesia/rabbit@HOST1/msg_stores/vhosts/628WB79CIFDYO9LJI6DKMI09L/msg_store_persistent'
2021-07-27 10:46:04.428 [info] <0.602.0> Message store for directory '/var/lib/rabbitmq/mnesia/rabbit@HOST1/msg_stores/vhosts/557K2FATBOJGJWHPLEDCZ3CRE/msg_store_persistent' is stopped
2021-07-27 10:46:04.428 [info] <0.598.0> Stopping message store for directory '/var/lib/rabbitmq/mnesia/rabbit@HOST1/msg_stores/vhosts/557K2FATBOJGJWHPLEDCZ3CRE/msg_store_transient'
2021-07-27 10:46:04.428 [info] <0.659.0> Message store for directory '/var/lib/rabbitmq/mnesia/rabbit@HOST1/msg_stores/vhosts/CVFDVHFSE3TE8DQ5RL2Z9Z7SM/msg_store_persistent' is stopped
2021-07-27 10:46:04.429 [info] <0.655.0> Stopping message store for directory '/var/lib/rabbitmq/mnesia/rabbit@HOST1/msg_stores/vhosts/CVFDVHFSE3TE8DQ5RL2Z9Z7SM/msg_store_transient'
2021-07-27 10:46:04.430 [info] <0.785.0> Stopping message store for directory '/var/lib/rabbitmq/mnesia/rabbit@HOST1/msg_stores/vhosts/B24YDZSK6ODW52OLV52U91PJQ/msg_store_persistent'
2021-07-27 10:46:04.432 [info] <0.716.0> Message store for directory '/var/lib/rabbitmq/mnesia/rabbit@HOST1/msg_stores/vhosts/628WB79CIFDYO9LJI6DKMI09L/msg_store_persistent' is stopped
2021-07-27 10:46:04.432 [info] <0.712.0> Stopping message store for directory '/var/lib/rabbitmq/mnesia/rabbit@HOST1/msg_stores/vhosts/628WB79CIFDYO9LJI6DKMI09L/msg_store_transient'
2021-07-27 10:46:04.433 [info] <0.598.0> Message store for directory '/var/lib/rabbitmq/mnesia/rabbit@HOST1/msg_stores/vhosts/557K2FATBOJGJWHPLEDCZ3CRE/msg_store_transient' is stopped
2021-07-27 10:46:04.433 [info] <0.655.0> Message store for directory '/var/lib/rabbitmq/mnesia/rabbit@HOST1/msg_stores/vhosts/CVFDVHFSE3TE8DQ5RL2Z9Z7SM/msg_store_transient' is stopped
2021-07-27 10:46:04.434 [info] <0.785.0> Message store for directory '/var/lib/rabbitmq/mnesia/rabbit@HOST1/msg_stores/vhosts/B24YDZSK6ODW52OLV52U91PJQ/msg_store_persistent' is stopped
2021-07-27 10:46:04.435 [info] <0.778.0> Stopping message store for directory '/var/lib/rabbitmq/mnesia/rabbit@HOST1/msg_stores/vhosts/B24YDZSK6ODW52OLV52U91PJQ/msg_store_transient'
2021-07-27 10:46:04.436 [info] <0.712.0> Message store for directory '/var/lib/rabbitmq/mnesia/rabbit@HOST1/msg_stores/vhosts/628WB79CIFDYO9LJI6DKMI09L/msg_store_transient' is stopped
2021-07-27 10:46:04.438 [info] <0.778.0> Message store for directory '/var/lib/rabbitmq/mnesia/rabbit@HOST1/msg_stores/vhosts/B24YDZSK6ODW52OLV52U91PJQ/msg_store_transient' is stopped
2021-07-27 10:46:04.445 [info] <0.44.0> Application rabbit exited with reason: stopped
2021-07-27 10:46:04.445 [info] <0.2986.0> Stopping application 'rabbitmq_prelaunch'
2021-07-27 10:46:04.445 [info] <0.44.0> Application rabbit exited with reason: stopped
2021-07-27 10:46:04.448 [info] <0.44.0> Application rabbitmq_prelaunch exited with reason: stopped
2021-07-27 10:46:04.448 [info] <0.2986.0> Stopping application 'rabbit_common'
2021-07-27 10:46:04.448 [info] <0.44.0> Application rabbitmq_prelaunch exited with reason: stopped
2021-07-27 10:46:04.448 [info] <0.2986.0> Stopping application 'sysmon_handler'
2021-07-27 10:46:04.448 [info] <0.44.0> Application rabbit_common exited with reason: stopped
2021-07-27 10:46:04.448 [info] <0.44.0> Application rabbit_common exited with reason: stopped
2021-07-27 10:46:04.450 [info] <0.2986.0> Stopping application 'os_mon'
2021-07-27 10:46:04.450 [info] <0.44.0> Application sysmon_handler exited with reason: stopped
2021-07-27 10:46:04.450 [info] <0.44.0> Application sysmon_handler exited with reason: stopped
2021-07-27 10:46:04.451 [info] <0.2986.0> Stopping application 'amqp10_client'
2021-07-27 10:46:04.451 [info] <0.44.0> Application os_mon exited with reason: stopped
2021-07-27 10:46:04.451 [info] <0.44.0> Application os_mon exited with reason: stopped
2021-07-27 10:46:04.453 [info] <0.44.0> Application amqp10_client exited with reason: stopped
2021-07-27 10:46:04.453 [info] <0.2986.0> Stopping application 'amqp10_common'
2021-07-27 10:46:04.453 [info] <0.2986.0> Stopping application 'ra'
2021-07-27 10:46:04.453 [info] <0.44.0> Application amqp10_client exited with reason: stopped
2021-07-27 10:46:04.453 [info] <0.44.0> Application amqp10_common exited with reason: stopped
2021-07-27 10:46:04.453 [info] <0.44.0> Application amqp10_common exited with reason: stopped
2021-07-27 10:46:04.455 [info] <0.44.0> Application ra exited with reason: stopped
2021-07-27 10:46:04.455 [info] <0.2986.0> Stopping application 'mnesia'
2021-07-27 10:46:04.455 [info] <0.44.0> Application ra exited with reason: stopped
2021-07-27 10:46:04.459 [info] <0.44.0> Application mnesia exited with reason: stopped
2021-07-27 10:46:04.459 [info] <0.2986.0> Successfully stopped RabbitMQ and its dependencies
2021-07-27 10:46:04.459 [info] <0.44.0> Application mnesia exited with reason: stopped
2021-07-27 10:46:04.459 [info] <0.2986.0> Halting Erlang VM with the following applications:
    lager
    observer_cli
    stdout_formatter
    gen_batch_server
    aten
    ranch
    cuttlefish
    inets
    credentials_obfuscation
    recon
    jsx
    goldrush
    xmerl
    tools
    syntax_tools
    ssl
    public_key
    asn1
    crypto
    compiler
    sasl
    stdlib
    kernel
```

Reply:
For the record (and search engines guiding users here), the discussion sub-discussion has moved here: https://github.com/rabbitmq/rabbitmq-server/discussions/3272